sui_programmability/framework/sources/TokenMetadata.move

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Metadata standard for fungible and non-fungible tokens on Sui
+module Sui::TokenMetadata {
+    use Sui::Url::{Self, Url};
+    use Sui::UTF8;
+    use Std::Vector;
+
+    struct TokenMetadata has store {
+        /// Name for the token
```

Comment:
I don't think we want a separate struct like this.

I think we want to allow any sort of composition for these fields. That is, you can add `name: String`, `description: String`, and `urls: vector<Url>` to your struct and it will get rendered properly. You can then include some, none, or all of these combos

I don't think we gain too much from having this separate struct? (other than maybe discoverability of these field combos)

Reply:
Yeah I thought about this but I do think it's worth having a separate struct with the following advantages:
1. less code

An application developer can write 3 lines of code instead of 9~12 lines of code(see inline comment). Moreover, imagine a module that has multiple NFT-like structs, They have to repeat `name, description, urls` in the struct definition and initializations for each struct. Chunking these fields in a standalone struct is much cleaner.

```
use Sui::TokenMetadata::{Self, TokenMetadata};
metadata: TokenMetadata,
metadata: TokenMetadata::new(name, description, url)
```


```
use Sui::Url::{Self, Url};
use Sui::UTF8;
use Std::Vector;

name: UTF8::String,
description: UTF8::String,
urls: vector<Url>,

name: UTF8::string_unsafe(name),
description: UTF8::string_unsafe(description),
// very likely we will need more when we support multiple urls at initialization
urls: Vector::singleton<Url>(Url::new_from_bytes_unsafe(url)),
```

2. Less error-prone

What if the developers do not read our documentation carefully, and ended up using `url` instead of `urls` or use `ASCII::String` instead of `UTF8::String` in the definition? Why should we require them to figure out the difference between `ASCII::String` and `UTF8::String` when all they want to do is pass some kind of string in a Hackathon?

3. Easier for application rendering

With the struct, the applications can just do `object.metadata` to render an object
Without the struct, the applications have to do `if 'name' in object; if 'description' in object; if 'urls' in object`.

These three fields are more commonly used together than separately, as evident in all other metadata standard(e.g., [metaplex](https://docs.metaplex.com/token-metadata/specification#token-standards), [opensea](https://docs.opensea.io/docs/metadata-standards)). Therefore putting them in a struct makes sense instead of allowing a combo. If the application wants to opt-out, they can pass in an empty string/vector.

Reply:
For ASCII/UTF8 string, we will support both

For url/urls we will support both (but we might also have linters), (we are supporting both for the reason of many vs multiple)

> Easier for application rendering
> An application developer can write 3 lines of code instead of 9~12 lines of code(see inline comment). Moreover, imagine a module that has multiple NFT-like structs, They have to repeat name, description, urls in the struct definition and initializations for each struct

My worry for both of these is we are going to have more meta data fields, other than just these 3.  We are going to have a whole bag of optional "meta data" fields, and we want some generic way of annotating meta data fields vs attribute fields. I'm not quite sure this struct approach is flexible for an extensible API.

Reply:
Got it, I was thinking having all the essential fields(most likely just these 3) in this `TokenMetada` struct, and put all the optional ones inline with the NFT definition(instead of in the `TokenMetadata`). That way we can have the best of both world and don't need to force people to write 9~12 lines of code or more instead of 3 lines of code.

> For ASCII/UTF8 string, we will support both

> For url/urls we will support both (but we might also have linters), (we are supporting both for the reason of many vs multiple)

It's not just if we decide to support or not. This will also put a burden on applications where they now need to do `if 'urls' in object; else if 'url' in object`. what if the developer use `title` instead of `name`?

Reply:
> I was thinking having all the essential fields(most likely just these 3) in this TokenMetada struct, and put all the optional ones inline with the NFT definition(instead of in the TokenMetadata). 

We could do that sure. I just worry about the complexity it will add for tooling. I don't think having the struct is that much more ergonomic for the programmer than the 3 fields, as it adds a layer of indirection when packing(creating), unpacking(pattern matching), reading, writing (mabye that last one won't happen though). 

It also makes the story around the field+type name thing a lot more complicated. 

My hope would be something simple like:
- Known field+type combos are specially
- For fields not known by the explorer
  - Any field with a leading `_`, e.g. `_foo: u64`,  is a meta data field
  - All other fields are "attribute" or "property" fields

Reply:
> It's not just if we decide to support or not. This will also put a burden on applications where they now need to do if 'urls' in object; else if 'url' in object.

This is a very good point. But I don't think it will just be solved by the `TokenMetadata` case. 
Specifically if end up with cases such as `url: Url` vs `url: VerifiedUrl` where `VerifiedUrl` is a Url + hash (@kchalkias was talking about this maybe being necessary)

At any rate, it might be helpful to have a tool that takes the JSON for a Sui Move object and canonicalizes it based on the rules/format the Sui explorer expects (bucketing these major fields like name/description vs attributes vs misc. metadata)